// @cosmos-kit/core/cjs/iframe/extension/client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IframeClient = void 0;
const types_1 = require("../../types");
const registry_1 = require("./registry");
const signers_1 = require("./signers");
const utils_1 = require("./utils");
class IframeClient {
    wallet;
    constructor(wallet) {
        this.wallet = wallet;
    }
    async getSimpleAccount(...params) {
        return await (0, utils_1.sendAndListenOnce)({
            method: 'getSimpleAccount',
            params,
        }, async (data) => {
            if (data.type === 'success') {
                return data.response;
            }
            else {
                throw new Error(data.error);
            }
        });
    }
    async connect(...params) {
        await (0, utils_1.sendAndListenOnce)({
            method: 'connect',
            params,
        }, async (data) => {
            if (data.type === 'success') {
                // On connect, update info based on data from parent.
                registry_1.iframeExtensionInfo.prettyName = data.response.prettyName;
                registry_1.iframeExtensionInfo.logo = data.response.logo;
            }
            else if (data.type === 'error') {
                if (data.error === types_1.IFRAME_PARENT_DISCONNECTED) {
                    // On disconnect, remove parent info.
                    registry_1.iframeExtensionInfo.prettyName = types_1.IFRAME_DEFAULT_PRETTY_NAME;
                    registry_1.iframeExtensionInfo.logo = types_1.IFRAME_DEFAULT_LOGO;
                    await this.wallet.disconnect();
                }
                throw new Error(data.error);
            }
        });
    }
    // Cannot implement because the event only makes sense in the context of the
    // DOM where it happened, and references to the iframe's window are not
    // available in the parent.
    // on(_type: string, _listener: EventListenerOrEventListenerObject) {
    //   throw new Error('Not implemented');
    // }
    // off(_type: string, _listener: EventListenerOrEventListenerObject) {
    //   throw new Error('Not implemented');
    // }
    async enable(...params) {
        await (0, utils_1.sendAndListenOnce)({
            method: 'enable',
            params,
        }, async (data) => {
            if (data.type === 'error') {
                throw new Error(data.error);
            }
        });
    }
    async suggestToken(...params) {
        await (0, utils_1.sendAndListenOnce)({
            method: 'suggestToken',
            params,
        }, async (data) => {
            if (data.type === 'error') {
                throw new Error(data.error);
            }
        });
    }
    async addChain(...params) {
        // `clientOptions` in the chain record may contain
        // `SigningStargateClientOptions` or `SigningCosmWasmClientOptions`, which
        // may contain a types registry that cannot be cloned when using
        // window.postMessage. `addChain` functions typically do not care about
        // `clientOptions` as they are just used for adding chain metadata to a
        // wallet, so we remove the `clientOptions` from the chain record here.
        let chainRecord = params[0];
        if (chainRecord && 'clientOptions' in chainRecord) {
            chainRecord = {
                ...chainRecord,
                clientOptions: {},
            };
            params[0] = chainRecord;
        }
        await (0, utils_1.sendAndListenOnce)({
            method: 'addChain',
            params,
        }, async (data) => {
            if (data.type === 'success') {
                return data.response;
            }
            else {
                throw new Error(data.error);
            }
        });
    }
    async getAccount(...params) {
        return await (0, utils_1.sendAndListenOnce)({
            method: 'getAccount',
            params,
        }, async (data) => {
            if (data.type === 'success') {
                return data.response;
            }
            else {
                throw new Error(data.error);
            }
        });
    }
    getOfflineSigner(chainId, preferredSignType) {
        if (preferredSignType === 'direct') {
            return this.getOfflineSignerDirect(chainId);
        }
        else {
            return this.getOfflineSignerAmino(chainId);
        }
    }
    getOfflineSignerAmino(chainId) {
        return new signers_1.IframeAminoSigner(chainId);
    }
    getOfflineSignerDirect(chainId) {
        return new signers_1.IframeDirectSigner(chainId);
    }
    async sign(...params) {
        return await (0, utils_1.sendAndListenOnce)({
            method: 'sign',
            params,
        }, async (data) => {
            if (data.type === 'success') {
                return data.response;
            }
            else {
                throw new Error(data.error);
            }
        });
    }
    async signAmino(...params) {
        return await (0, utils_1.sendAndListenOnce)({
            method: 'signAmino',
            params,
        }, async (data) => {
            if (data.type === 'success') {
                return data.response;
            }
            else {
                throw new Error(data.error);
            }
        });
    }
    async signDirect(...params) {
        return await (0, utils_1.sendAndListenOnce)({
            method: 'signDirect',
            params,
        }, async (data) => {
            if (data.type === 'success') {
                return data.response;
            }
            else {
                throw new Error(data.error);
            }
        });
    }
    async signArbitrary(...params) {
        return await (0, utils_1.sendAndListenOnce)({
            method: 'signArbitrary',
            params,
        }, async (data) => {
            if (data.type === 'success') {
                return data.response;
            }
            else {
                throw new Error(data.error);
            }
        });
    }
    async getEnigmaPubKey(...params) {
        return await (0, utils_1.sendAndListenOnce)({
            method: 'getEnigmaPubKey',
            params,
        }, async (data) => {
            if (data.type === 'success') {
                return data.response;
            }
            else {
                throw new Error(data.error);
            }
        });
    }
    async getEnigmaTxEncryptionKey(...params) {
        return await (0, utils_1.sendAndListenOnce)({
            method: 'getEnigmaTxEncryptionKey',
            params,
        }, async (data) => {
            if (data.type === 'success') {
                return data.response;
            }
            else {
                throw new Error(data.error);
            }
        });
    }
    async enigmaEncrypt(...params) {
        return await (0, utils_1.sendAndListenOnce)({
            method: 'enigmaEncrypt',
            params,
        }, async (data) => {
            if (data.type === 'success') {
                return data.response;
            }
            else {
                throw new Error(data.error);
            }
        });
    }
    async enigmaDecrypt(...params) {
        return await (0, utils_1.sendAndListenOnce)({
            method: 'enigmaDecrypt',
            params,
        }, async (data) => {
            if (data.type === 'success') {
                return data.response;
            }
            else {
                throw new Error(data.error);
            }
        });
    }
    async sendTx(...params) {
        return await (0, utils_1.sendAndListenOnce)({
            method: 'sendTx',
            params,
        }, async (data) => {
            if (data.type === 'success') {
                return data.response;
            }
            else {
                throw new Error(data.error);
            }
        });
    }
}
exports.IframeClient = IframeClient;
